include/zephyr/drivers/sensor_v2.h

```diff
@@ -0,0 +1,989 @@
+ * @param offset An offset into the available sample rates to get.
+ * @param max_count The maximum number of sample rates to fetch
+ * @return 0 if no sample rates are available at the given offset
+ * @return > 0 if available sample rates were written to the sample_rates argument
+ * @return < 0 on error
+
+ */
+__syscall int sensor_get_sample_rate_available(const struct device *sensor, uint32_t sensor_type,
+					       uint32_t *sample_rates, uint8_t offset,
+					       uint8_t max_count);
+
+static inline int z_impl_sensor_get_sample_rate_available(const struct device *sensor,
+							  uint32_t sensor_type,
+							  uint32_t *sample_rates, uint8_t offset,
+							  uint8_t max_count)
+{
+#ifdef CONFIG_SENSOR_STREAMING_MODE
+	const struct sensor_driver_api_v2 *api = sensor->api;
+
+	if (api == NULL || api->get_sample_rate_available == NULL) {
+		return -ENOSYS;
+	}
+
+	return api->get_sample_rate_available(sensor, sensor_type, sample_rates, offset, max_count);
+#else
+	return -ENOSYS;
+#endif
+}
+
+/**
+ * @brief Get the sensor's sample rate
+ *
+ * Get the sample rate of the sensor in milli-Hz.
+ *
+ * @param sensor Pointer to the sensor device
+ * @param sensor_type The type of sensor to read
+ * @param sample_rate Pointer to the sample rate variable that will be set on
+ *        success
+ * @return 0 on success
+ * @return
+ */
+__syscall int sensor_get_sample_rate(const struct device *sensor, uint32_t sensor_type,
+				     uint32_t *sample_rate);
+
+static inline int z_impl_sensor_get_sample_rate(const struct device *sensor, uint32_t sensor_type,
+						uint32_t *sample_rate)
+{
+#ifdef CONFIG_SENSOR_STREAMING_MODE
+	const struct sensor_driver_api_v2 *api = sensor->api;
+
+	if (api == NULL || api->get_sample_rate == NULL) {
+		return -ENOSYS;
+	}
+
+	return api->get_sample_rate(sensor, sensor_type, sample_rate);
+#else
+	return -ENOSYS;
+#endif
+}
+
+/**
+ * @brief Update the sensor's sample rate
+ *
+ * Set a new sample rate for the sensor. Sample rates should be in milli-Hz.
+ *
+ * @param sensor Pointer to the sensor device
+ * @param sensor_type The type of sensor to read
+ * @param sample_rate The new sample rate in milli-Hz
+ * @param round_up Whether or not to round up to the nearest available sample rate
+ * @return 0 on success
+ * @return < 0 on failure
+ */
+__syscall int sensor_set_sample_rate(const struct device *sensor, uint32_t sensor_type,
+				     uint32_t sample_rate, bool round_up);
+
+static inline int z_impl_sensor_set_sample_rate(const struct device *sensor, uint32_t sensor_type,
+						uint32_t sample_rate, bool round_up)
+{
+#ifdef CONFIG_SENSOR_STREAMING_MODE
+	const struct sensor_driver_api_v2 *api = sensor->api;
+
+	if (api == NULL || api->set_sample_rate == NULL) {
+		return -ENOSYS;
+	}
+
+	return api->set_sample_rate(sensor, sensor_type, sample_rate, round_up);
+#else
+	return -ENOSYS;
+#endif
+}
+/**
+ * @brief Get the sensor FIFO watermark value as a percent
```

Comment:
Is "watermark" specific to the BMI160 or a generally used term? I am familiar with "high water mark" with similar meaning, but "watermark" sounds like some kind of unique identifier that gets added to each of the reading values. How would you feel about something like, "interrupt threshold"?

Reply:
Watermark is pretty widely used from what I've seen... tdk, bosch, kionix, nxp, and st sensors use the term at least.

Reply:
Standard term for me 👍